design/20210203.certificate-request-identity.md

```diff
@@ -0,0 +1,165 @@
+    identity in a useful way
+
+### Non-Goals
+
+- Discuss how this identity can be used (policy, auditing)
+- Make changes to upstream Kubernetes to implement identiy in cert-manager
+- Although considered below, "passed down" identity is not part of this design
+
+## Proposal
+
+Kubernetes does [not currently support immutable
+fields](https://github.com/kubernetes/enhancements/blob/8b9b994136371f1bc938aabf012f4c45535d684c/keps/sig-api-machinery/20190603-immutable-fields.md)
+for CRDs. The cert-manager webhook will be responsible for populating and
+enforcing identity fields which are present on the spec of `CertificateRequest`
+resources.
+
+The webhook will be responsible for enforcing the following during a CREATE
+operation:
+- No identity fields have been set by the user creating the resource
+- Set identity fields to exactly what is received from the API server in the
+    [UserInfo](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#userinfo-v1-authentication-k8s-io)
+
+The webhook will also responsible for enforcing the following during an UPDATE
+operation:
+- No changes to identity fields are allowed to be made
+
+### API Changes
+
+In order to expose the identity of who created `CertificateRequest` resources,
+these resources must be updated to have parity with the upstream Kubernetes
+certificates API.
+This means that the `CertificateRequest` API type be updated to include the
+following fields in spec, for all API versions:
+
+```go
+	// Username contains the name of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	Username string `json:"username,omitempty"`
+	// UID contains the uid of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	UID string `json:"uid,omitempty"`
+	// Groups contains group membership of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +listType=atomic
+	// +optional
+	Groups []string `json:"groups,omitempty"`
+	// Extra contains extra attributes of the user that created the CertificateRequest.
+	// Populated by the cert-manager webhook on creation and immutable.
+	// +optional
+	Extra map[string][]string `json:"extra,omitempty"`
+```
+
+### Upgrading
+
+Any `CertificateRequest`s that are present in the cluster at the time of
+upgrade, won't and will never have their identity fields populated. This should
+be acceptable.
+
+All subsequent `CertificateRequest`s created after this upgrade will have their
+identities populated.
+
+### Test Plan
+
+Unit tests will be created is ensure that the [properties](#proposal) are
+enforced.
+
+End to end tests will ensure that different users creating `CertificateRequest`
+resources have their respective identities populated in the fields.
+
+### Risks and Mitigations
```

Comment:
Hm, just another thought... if someone were to configure their own mutating webhook, there could potentially be cases where the other webhook *overrides* the UserInfo data with other, incorrect data (e.g. maliciously).

I think we need to verify in the *validation webhook* that the UserInfo of the request also matches the UserInfo in the body of the resource *iff* any of the fields in the body of the resource are set (and this is a create operation)

Reply:
Yeah that's what I have done, see above ^

Should be called out in the design doc though I think

Reply:
https://github.com/jetstack/cert-manager/blob/33f3fe473caaa38565690712ec9582919f335f51/pkg/internal/apis/certmanager/identity/certificaterequests/certificaterequests.go